### ISSUE TITLE
Making built wheels available for install

### ISSUE THREAD

**Issue by AbdBarho** [facebookresearch/xformers#533]
# 🚀 Feature

## Motivation

After #523 #534, the wheels can be built, but are not available for install anywhere. But users want this #532 #473

## Pitch & Alternatives

There a couple of ways that I know of to achieve this:

### Upload to pypi

Probably the most obvious solution

* max file size for pypi is 60MB, or roughly 2-3 CUDA architectures per wheel
  * maybe it makes sense to have per architecture? sounds like an overkill to me
  * one can request more size from the pypi team, I am not sure if this solution scales
* how would `pip` find the correct wheel for a given torch & cuda version?
  * probably using an extra pip index just like pytorch does it, but it is not "trivial" since now we need to setup an external service to host index, and in that case:

### External hosting

Upload the built wheel to an external storage / hosting service (s3 & co.)

* no size limits
* additional costs
* now it is not about python or wheels anymore, there is infrastructure management & maintenance that needs to taken care of. 
* does not solve the question of: how would `pip` find the correct wheel for a given torch & cuda version?
  * the pip index still needs to be setup (and updated automatically by every release), extra complexity

### Upload to the release page on github

Similar to the above, but using github

* max file size is 2GB (which should be enough)
* no costs
* still does not answer the pip question
* download times are slower than others, but this should not be a problem

maybe it would be easiest to dump all wheels to github releases and then the users would have to find the version they want and install it.

`pip install https://github.com/..../xformers-1.14.0+torch1.12.1cu116-cp39-cp39-linux_x86_64.whl`

Although this would make it really annoying for libraries that use xformers and want to build for different python /cuda versions, you cannot just add `xformers` to your `requirements.txt`.

### Additional info & Resources

https://peps.python.org/pep-0503/

https://packaging.python.org/en/latest/guides/hosting-your-own-index/

I am willing to contribute if I know how.

**Comment by danthe3rd** [comment]
> * max file size for pypi is 60MB

Apparently it's now 100MB. I've requested an increase to 300MB which should give us some margin
https://github.com/pypa/pypi-support/issues/2421

> how would pip find the correct wheel for a given torch & cuda version?

That's the reason why we are currently using conda, which handles this sort of dependencies.
I see a few options:
(1) `pip install` manually with the full path from s3
(2) `pip install xformers` downloads a source distribution. At setup, we retrieve the pytorch/cuda/xformers/python/os version, and download the corresponding binary from s3 if available. Otherwise we build from source
(3) Pin a pytorch version in the requirements, and hope we have compatibility between different cuda versions (I've frequently mismatched 11.6/11.7 without issues..)

In any case, a follow-up task would be to check at run-time that the pytorch version has not changed


What do you think?

**Comment by AbdBarho** [comment]
Overall I think its a good plan, some comments:

> (2) pip install xformers downloads a source distribution. At setup, we retrieve the pytorch/cuda/xformers/python/os version, and download the corresponding binary from s3 if available. Otherwise we build from source

Could this in anyway lead to endless loops? I don't know how, but my spider sense is tingling.

Also, the order of installs here matters right? it won't not be possible to 
```sh
pip install torch xformers --extra-index-url https://download.pytorch.org/whl/cu116
```




> (3) Pin a pytorch version in the requirements, and hope we have compatibility between different cuda versions (I've frequently mismatched 11.6/11.7 without issues..)

same experience

Perhaps maybe a bit more concretely on the plan forward:

1. Wait for pypi team and add pip upload to the pipeline
    * what if we don't get the storage increase? 
2. Upload wheels to S3
    * how is the lifetime of the bucket managed? in this repo or somewhere else?
    * is this point mutually exclusive with point 1?

**Comment by danthe3rd** [comment]
> Also, the order of installs here matters right? it won't not be possible to

I think we can have pytorch an install dependency or something like that.
The thing is: we already import & use torch in the `setup.py` script (https://github.com/facebookresearch/xformers/blob/main/setup.py#L20)

> Wait for pypi team and add pip upload to the pipeline

This does not solve the issue with multiple versions of pytorch/cuda (unless we ignore cuda mismatch errors, and add a pinned version of pytorch - latest? - as a dependency)

**Comment by danthe3rd** [comment]
> Could this in anyway lead to endless loops? I don't know how, but my spider sense is tingling.

Oh I meant just download the built binary files (not the entire package). Like `_C.so` for linux

**Comment by AbdBarho** [comment]
@danthe3rd now that we have pypi storage, what is the next step? is there something I can help with? do we still need to worry about the pip index? it is just an html page after all but still.

**Comment by danthe3rd** [comment]
I'm trying to figure out internally how we can store that on s3 - it's thanksgiving in the US so hopefully we will have more news to share next week.

**Comment by danthe3rd** [comment]
It looks a bit more complicated than I expected - for now I think we could upload to [pypi](https://pypi.org/project/xformers/) directly.
I think a first step would be to upload files corresponding to the latest pytorch version on pypi (eg pytorch 1.13 with cu116) for win/linux. It looks like we need a few things:
(1) We need to build on "manylinux" rather than ubuntu (at least for linux) - or we might just hack it for now
(2) We should also change the version for the builds we upload, as pypi won't support local prefixes (eg +git{HASH}.d{DATE} that we use currently). We could add a ".dev" suffix with the number of commits in the branch, [like we do for conda builds](https://github.com/facebookresearch/xformers/blob/main/packaging/conda/build_conda.py#L94)

What do you think?
I can setup the secrets for twine once this is working :)

**Comment by AbdBarho** [comment]
for 1) we are already building on manylinux
https://github.com/facebookresearch/xformers/blob/1515f775739d89628f0293aadcdc3a820d43f287/.github/workflows/wheels.yml#L65
but the generated wheel has the suffix `linux_x64_86`, maybe we can cheese it and `sed` to `manylinux`, I used the 2014 version for python version support https://github.com/pypa/manylinux#manylinux

for 2) I always thought this was handled by `pip` automatically if the last commit is tagged, this requires some testing.

And for 3) an additional source build should be easy, since we don't need cuda / torch / compilation, unless I am missing something.


I have never uploaded to pip before, so I will have to play around first to get a feeling for it....

**Comment by AbdBarho** [comment]
@danthe3rd yeah, I will start with a dummy repo and https://test.pypi.org/, also check the versioning thing, hopefully over the weekend.

**Comment by danthe3rd** [comment]
That is right. @fmassa how is pypi upload handled in torchvision?
I assume we should be able to rely on NOVA to create the builds and then upload however we want.

Otherwise, we can also upload manually with twine if you want to give it a shot with a test repo.

**Comment by AbdBarho** [comment]
@danthe3rd I managed to get NOVA running:

https://github.com/AbdBarho/xformers-wheels/actions/runs/3655705207

I will clean up the code and create a PR.

pypi could be added easily from the generated artifacts

**Comment by danthe3rd** [comment]
Oh wow that was fast :o
Thanks a lot for doing this! If you happen to have any issue with NOVA, let me know and we can have them update their scripts :) (I was wondering if they have the pagefile size increase working for windows build to prevent OOM...)

**Comment by danthe3rd** [comment]
Oh nice! Let's only upload for the latest pytorch stable (1.13 at the moment), with cuda 11.7 for instance.

**Comment by FrancescoSaverioZuppichini** [comment]
Thanks for working on this :pray: 

I've downloaded the whl file `formers-0.0.15.dev383-cp310-cp310-manylinux2014_x86_64.whl `, I am on amd CPU so I guess I have to use an Intel one right?

```
pip install formers-0.0.15.dev383-cp310-cp310-manylinux2014_x86_64.whl 
ERROR: formers-0.0.15.dev383-cp310-cp310-manylinux2014_x86_64.whl is not a supported wheel on this platform.
```

**Comment by AbdBarho** [comment]
@danthe3rd That's is one question answered, which is the same as what I had in mind:
https://github.com/AbdBarho/xformers-wheels/blob/fde92dc29741be208431c9d1dd0ed365129066c9/.github/workflows/wheels.yml#L148

The second question is regarding the version, do we want to upload `dev` versions? or only tagged commits? and how should the version string look like?

I saw the conda example, the generated version string is not accepted by pypi, what I have now is just `dev` + num commits from latest tag.

**Comment by AbdBarho** [comment]
@FrancescoSaverioZuppichini oh oh! can you give me some details about your os? the wheel should (in theory) work on any linux machine, if it doesn't, we have a problem.

**Comment by FrancescoSaverioZuppichini** [comment]
@AbdBarho Sure, from `neofetch`

![image](https://user-images.githubusercontent.com/15908060/207364034-3bcfadba-edd1-49cb-ad83-3a015dcc734a.png)

**Comment by danthe3rd** [comment]
> The second question is regarding the version, do we want to upload `dev` versions? or only tagged commits? and how should the version string look like?

The optimal solution I see would be:
(1) Upload a "pre-release" for next version at every commit on main
(2) Upload a tagged release when we tag a version-like branch
We can start with whatever is easier for you

> I saw the conda example, the generated version string is not accepted by pypi, what I have now is just `dev` + num commits from latest tag.

Yes, pypi won't accept local suffixes. Let's remove them from the version number.

**Comment by AbdBarho** [comment]
@FrancescoSaverioZuppichini I managed to install the wheel on a debian and centos system.

This wheel is for python 3.10, can you try again with 
```
python3.10 -m pip install ./formers-0.0.15.dev383-cp310-cp310-manylinux2014_x86_64.whl
```
?

**Comment by AbdBarho** [comment]
@danthe3rd I guess we can start with the pre-release. tagged commits can come laterr.

**Comment by AbdBarho** [comment]
@FrancescoSaverioZuppichini yes yes, I am working on wheels for 3.7 ---- 3.10, Linux / windows, cuda only. (For now)

**Comment by danthe3rd** [comment]
And we just have our first binary wheel uploaded!
https://pypi.org/project/xformers/0.0.16rc377/#files

You can test it with `pip install -U --pre xformers`

**Comment by JacobAsmuth** [comment]
Thanks Dan! I really appreciate your hard work! 

After installing xformers with

`pip install -U --pre xformers`

and starting up the stable diffusion 2.0 model (run through the [stable-diffusion-ui](https://github.com/cmdr2/stable-diffusion-ui), I get the following error:

`RuntimeError: No such operator xformers::efficient_attention_forward_cutlass - did you forget to build xformers with 'python setup.py develop'?`

**Comment by AbdBarho** [comment]
@JacobAsmuth are you running pytorch 1.12.x?

**Comment by JacobAsmuth** [comment]
I am not, I'm running pytorch 1.11.0 .

**Comment by AbdBarho** [comment]
Unfortunately, The wheels are built against the latest pytorch 1.13 :/
We have some wheels for 1.12 [here](https://github.com/facebookresearch/xformers/actions/runs/3698361241), but nothing for 1.11.

Maybe @danthe3rd has any ideas?

**Comment by JacobAsmuth** [comment]
I'm using 1.11 only because that's what it appears that the https://github.com/Stability-AI/stablediffusion library installs with their default requirments.txt - let me see if it's compatible with later versions.

**Comment by danthe3rd** [comment]
Did we miss something @AbdBarho ? pip installing XFormers should have updated PyTorch as it's a requirement to have version==1.13...

We definitely need a better error message btw.

**Comment by AbdBarho** [comment]
Not sure, checking the METADATA of the wheel shows that it looks okay:
```
Metadata-Version: 2.1
Name: xformers
Version: 0.0.16rc379
Summary: XFormers: A collection of composable Transformer building blocks.
Home-page: https://facebookresearch.github.io/xformers/
Author: Facebook AI Research
Author-email: oncall+xformers@xmail.facebook.com
Classifier: Programming Language :: Python :: 3.7
Classifier: Programming Language :: Python :: 3.8
Classifier: Programming Language :: Python :: 3.9
Classifier: Programming Language :: Python :: 3.10
Classifier: License :: OSI Approved :: BSD License
Classifier: Topic :: Scientific/Engineering :: Artificial Intelligence
Classifier: Operating System :: OS Independent
Requires-Python: >=3.7
Description-Content-Type: text/markdown
License-File: LICENSE
Requires-Dist: numpy
Requires-Dist: pyre-extensions (==0.0.23)
Requires-Dist: einops
Requires-Dist: torch (==1.13)

XFormers: A collection of composable Transformer building blocks.XFormers aims at being able to reproduce most architectures in the Transformer-family SOTA,defined as compatible and combined building blocks as opposed to monolithic models

```

and the same applies to the source distribution.

Maybe because it is a pre-release?

**Comment by netw0rkf10w** [comment]
@danthe3rd @AbdBarho Does this work for torch 1.14 (aka 2.0) please?

**Comment by debdip** [comment]
It does work on torch 1.13. 
I just installed it.

![image](https://user-images.githubusercontent.com/11462610/207811793-18d7a9ff-ba45-419c-bf94-eb5be2e50041.png)

**Comment by netw0rkf10w** [comment]
> @danthe3rd @AbdBarho Does this work for torch 1.14 (aka 2.0) please?

Would it be possible to relax the `torch==1.13` constraint to `torch>=1.13` for the wheels?

**Comment by danthe3rd** [comment]
> Would it be possible to relax the `torch==1.13` constraint to `torch>=1.13` for the wheels?

xFormers works with any version of pytorch if you build yourself, but for the pre-built binary wheels, they are only compatible with the pytorch version they were built for. For now we chose to stick to the latest stable

**Comment by netw0rkf10w** [comment]
@danthe3rd Sorry I thought that `pip install` was building the package. I'll just install from source then. Thanks.

**Comment by danthe3rd** [comment]
@AbdBarho actually looking at how releases are tagged in pytorch (and previous releases in xformers), we will stick with tags like `vA.B.C`. Do you think you can update the script for that?
I've also added a todo-list to track release of 0.0.16 - hopefully this week or next week

**Comment by danthe3rd** [comment]
Hum it also looks like we have issues with the RC releases now: https://github.com/facebookresearch/xformers/actions/runs/3703118568/jobs/6274182582#step:10:21
Maybe we should use the context for the github action somewhere; https://docs.github.com/en/actions/learn-github-actions/contexts#job-context

**Comment by FrancescoSaverioZuppichini** [comment]
@AbdBarho  it works! :rocket: Thanks a lot :heart: 

Tested on `python 3.9/3.10`

**Comment by FrancescoSaverioZuppichini** [comment]
@AbdBarho wait ... you are the hero who make https://github.com/AbdBarho/stable-diffusion-webui-docker sending all the love of this world man! 🤗

**Comment by pcuenca** [comment]
This is awesome!

We (Hugging Face [diffusers](https://github.com/huggingface/diffusers) team) really want to recommend users to install xFormers, and it'll be much easier when pip wheels are available.

My understanding from #591 is that wheels will be automatically made available via `pip` starting from version `0.0.16` (for PyTorch 1.13). Is that correct?

**Comment by danthe3rd** [comment]
Yes that is correct! We will also pin the associated PyTorch version, as only a single pytorch can be supported per pypi version for binary wheels

**Comment by pcuenca** [comment]
> Yes that is correct! We will also pin the associated PyTorch version, as only a single pytorch can be supported per pypi version for binary wheels

Excellent, we'll update our documentation when `0.0.16` is released. Thanks a lot!

**Comment by danthe3rd** [comment]
@AbdBarho FYI we changed the version for the pip wheels to match the conda versions: https://github.com/fairinternal/xformers/commit/c4092d0cd6c2d43be875c35b597ab32af0a96067

**Comment by AbdBarho** [comment]
Can we close this now? it seems that everything is working fine.

**Comment by danthe3rd** [comment]
Yes indeed, good point.
Thanks again for all your contributions!
